### PR TITLE
[Dot-3923] feat: implement fetch results feature

### PR DESCRIPTION
**Pull request by parthlambdatest**

### CODE REVIEW

File: src/lib/httpClient.ts [hunk 1]
```diff
@@ -64,6 +64,22 @@ export default class httpClient {
         }, log)
     }
 
+    getScreenshotData(buildId: string, baseline: boolean, log: Logger) {
+        return this.request({
+            url: '/getScreenshotData',
```
Comment (sushobhit-lt) [inline src/lib/httpClient.ts:69]: Changes the route to `/screenshot`

File: src/lib/httpClient.ts [hunk 2]
```diff
@@ -64,6 +64,22 @@ export default class httpClient {
...
+            method: 'GET',
+            params: { buildId, baseline }
+        }, log)
+            .then(response => {
+                log.debug(`Fetched screenshot data successfully for buildId: ${buildId}`);
+                return response;
+            })
+            .catch(error => {
+                log.error(`Error fetching screenshot data for buildId: ${buildId}`);
```
Comment (sushobhit-lt) [inline src/lib/httpClient.ts:78]: log error as well

File: src/lib/httpClient.ts (hunk 1 above)
Comment (parthlambdatest) [reply src/lib/httpClient.ts:69]: We already have a route with /screenshot in LSRS, tough its a POST method.

File: src/lib/utils.ts [hunk 3]
```diff
@@ -199,4 +201,79 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
+                    }
+                }
+
+                // Display summary
+                console.log(`\nSummary of Mismatches:`);
+                console.log(`Total Variants with Mismatches: ${totalVariantsWithMismatches} out of ${totalVariants}`);
+                console.log(`Total Screenshots with Mismatches: ${totalScreenshotsWithMismatches} out of ${totalScreenshots}`);
+                console.log(`Branch Name: ${resp.build.branch_name}`);
+                console.log(`Project Name: ${resp.project.name}`);
+                console.log(`Build ID: ${resp.build.build_id}`);
```
Comment (sushobhit-lt) [inline src/lib/utils.ts:271]: All above statements can be compressed into single log

File: src/lib/utils.ts [hunk 4]
```diff
@@ -199,4 +201,79 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
+            return;
+        }
+        
+        try {
+            const resp = await ctx.client.getScreenshotData(ctx.build.id, ctx.build.baseline, ctx.log);
+
+            fs.writeFileSync(ctx.options.fetchResultsFileName, JSON.stringify(resp, null, 2));
+            ctx.log.debug(`Updated results in ${ctx.options.fetchResultsFileName}`);
+
+            if (resp.build.build_status_ind === 'completed' || resp.build.build_status_ind === 'error') {
```
Comment (sushobhit-lt) [inline src/lib/utils.ts:234]: add check for `resp.build` is null 

Use an Array of Constants for 'completed','error' status, there might be will add new status as well

File: src/commander/capture.ts [hunk 5]
```diff
@@ -17,9 +17,18 @@ command
     .description('Capture screenshots of static sites')
     .argument('<file>', 'Web static config file')
     .option('--parallel', 'Capture parallely on all browsers')
+    .option('--fetch-results [filename]', 'Fetch results and optionally specify an output file, e.g., <filename>.json')
     .action(async function(file, _, command) {
         let ctx: Context = ctxInit(command.optsWithGlobals());
 
+        const { fetchResults } = command.opts();
```
Comment (pinanks) [inline src/commander/capture.ts:24]: Do this part in ctxInit

File: src/commander/exec.ts [hunk 6]
```diff
@@ -19,9 +19,18 @@ command
     .description('Run test commands around SmartUI')
     .argument('<command...>', 'Command supplied for running tests')
     .option('-P, --port <number>', 'Port number for the server')
+    .option('--fetch-results [filename]', 'Fetch results and optionally specify an output file, e.g., <filename>.json')
     .action(async function(execCommand, _, command) {
         let ctx: Context = ctxInit(command.optsWithGlobals());
 
+        const { fetchResults } = command.opts();
```
Comment (pinanks) [inline src/commander/exec.ts:26]: Same. Have this in ctxInit

File: src/commander/upload.ts [hunk 7]
```diff
@@ -39,6 +40,14 @@ command
         }
         ctx.uploadFilePath = directory;
 
+        const { fetchResults } = command.opts();
```
Comment (pinanks) [inline src/commander/upload.ts:43]: ctxInit. This piece of code is repeated.

File: src/commander/uploadFigma.ts [hunk 8]
```diff
@@ -19,9 +19,18 @@ command
     .argument('<file>', 'figma design config file')
     .option('--markBaseline', 'Mark the uploaded images as baseline')
     .option('--buildName <buildName>' , 'Name of the build')
+    .option('--fetch-results [filename]', 'Fetch results and optionally specify an output file, e.g., <filename>.json')
     .action(async function(file, _, command) {
         let ctx: Context = ctxInit(command.optsWithGlobals());
 
+        const { fetchResults } = command.opts();
```
Comment (pinanks) [inline src/commander/uploadFigma.ts:26]: ctxInit

File: src/lib/utils.ts [hunk 9]
```diff
@@ -199,4 +201,79 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
     ];
+}
+
+// Global SIGINT handler
+process.on('SIGINT', () => {
+    if (isPollingActive) {
+        console.log('Fetching results interrupted. Exiting...');
+        isPollingActive = false;
+    } else {
+        console.log('\nExiting gracefully...');
```
Comment (pinanks) [inline src/lib/utils.ts:212]: Don't use console.log() anywhere. Use the winston logger.

File: src/lib/utils.ts [hunk 10]
```diff
@@ -199,4 +201,79 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
+                // Display summary
+                console.log(`\nSummary of Mismatches:`);
+                console.log(`Total Variants with Mismatches: ${totalVariantsWithMismatches} out of ${totalVariants}`);
+                console.log(`Total Screenshots with Mismatches: ${totalScreenshotsWithMismatches} out of ${totalScreenshots}`);
+                console.log(`Branch Name: ${resp.build.branch_name}`);
+                console.log(`Project Name: ${resp.project.name}`);
+                console.log(`Build ID: ${resp.build.build_id}`);
+            }
+        } catch (error: any) {
+            console.error(`Error fetching screenshot data: ${error.message}`);
```
Comment (pinanks) [inline src/lib/utils.ts:274]: Use winston logger

File: src/lib/utils.ts (hunk 3 above)
Comment (pinanks) [reply src/lib/utils.ts:271]: Use task output here.

**Review by pinanks — APPROVED** [review]

File: src/lib/httpClient.ts [hunk 11]
```diff
@@ -64,6 +64,22 @@ export default class httpClient {
         }, log)
     }
 
+    getScreenshotData(buildId: string, baseline: boolean, log: Logger) {
+        return this.request({
+            url: '/screenshot',
+            method: 'GET',
+            params: { buildId, baseline }
+        }, log)
+            .then(response => {
```
Comment (pinanks) [inline src/lib/httpClient.ts:73]: No need for then catch and logging here, request function handles that already.

File: src/lib/utils.ts [hunk 12]
```diff
@@ -199,4 +201,87 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
         ...mobileRenderViewports[constants.MOBILE_OS_IOS],
         ...mobileRenderViewports[constants.MOBILE_OS_ANDROID]
     ];
+}
+
+// Global SIGINT handler
+process.on('SIGINT', () => {
+    if (isPollingActive) {
+        console.log('Fetching results interrupted. Exiting...');
```
Comment (pinanks) [inline src/lib/utils.ts:209]: Use ctx.log

File: src/lib/utils.ts [hunk 13]
```diff
@@ -199,4 +201,87 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
+        isPollingActive = false;
+    } else {
+        console.log('\nExiting gracefully...');
```
Comment (pinanks) [inline src/lib/utils.ts:212]: Use ctx.log

File: src/lib/utils.ts [hunk 14]
```diff
@@ -199,4 +201,87 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
+        if (!isPollingActive) {
+            clearInterval(intervalId);
+            return;
+        }
+        
+        try {
+            const resp = await ctx.client.getScreenshotData(ctx.build.id, ctx.build.baseline, ctx.log);
+
+            if (!resp.build) {
+                console.error("Error: Build data is null.");
```
Comment (pinanks) [inline src/lib/utils.ts:232]: Use ctx.log

File: src/lib/utils.ts [hunk 15]
```diff
@@ -199,4 +201,87 @@ export function getRenderViewportsForOptions(options: any): Array<Record<string,
...
+                    `\nSummary of Mismatches:\n` +
+                    `Total Variants with Mismatches: ${totalVariantsWithMismatches} out of ${totalVariants}\n` +
+                    `Total Screenshots with Mismatches: ${totalScreenshotsWithMismatches} out of ${totalScreenshots}\n` +
+                    `Branch Name: ${resp.build.branch}\n` +
+                    `Project Name: ${resp.project.name}\n` +
+                    `Build ID: ${resp.build.build_id}\n`
+                );                
+            }
+        } catch (error: any) {
+            console.error(`Error fetching screenshot data: ${error.message}`);
```
Comment (pinanks) [inline src/lib/utils.ts:282]: Use ctx.log

File: src/lib/utils.ts (hunk 12 above)
Comment (parthlambdatest) [reply src/lib/utils.ts:209]: Will address later if required

File: src/lib/utils.ts (hunk 13 above)
Comment (parthlambdatest) [reply src/lib/utils.ts:212]: Will address later if required

**Pull request merged**